### PR TITLE
refactor(components): Remove redundant button prop types

### PR DESCRIPTION
**Pull request by davidr64**
## Motivations
Simplify reading the button prop type so developers can more easily read the type to understand what it is trying to achieve.

## Changes

Removed unneeded inheritance in the sub types.
Removed unused destructive type
Separated type to give descriptive names

## Testing

Paste this code into the button file. All the objects added to the workingButtons array should be valid 
All the buttons in the badButtons array should have a typescript error
The label or aria label should give some explanation.
Do this on master and in the branch. Behaviour should be mostly the same except for the last bad button preventing the `external` flag existing on with the `to` property.
```
const workingButtons: ButtonProps[] = [];
workingButtons.push({
  label: "Button that doesn't do anything",
  //should we make it so onClick, submit, to, or url is required?
});

workingButtons.push({
  label: "nav button",
  to: "",
});

workingButtons.push({
  icon: "search",
  ariaLabel: "submit button must be primary",
  type: "primary",
  submit: true,
});

workingButtons.push({
  label: "url button",
  url: "",
});

workingButtons.push({
  label: "external url button",
  url: "",
  external: true,
});

workingButtons.push({
  label: "Destruction",
  onClick: () => {},
});

//new submit can be destructive
workingButtons.push({
  label: "Button can't submit and destructively",
  submit: true,
  variation: "destructive",
});

//done so there isn't an error on workingButtons not being used
console.log(workingButtons);

const badButtons: ButtonProps[] = [];
badButtons.push({
  label: "Button can't submit and onclick",
  onClick: () => {},
  submit: true,
});


workingButtons.push({
  icon: "search",
  ariaLabel: "submit button can't have variations",
  type: "primary",
  variation: "learning",
  submit: true,
});

badButtons.push({
  label: "Button can't submit and to",
  submit: true,
  to: "",
});

badButtons.push({
  label: "Button can't url and to",
  url: "",
  to: "",
});

badButtons.push({
  icon: "search",
  //needs a aria label
  type: "primary",
  submit: true,
});

badButtons.push({
  icon: "search",
  ariaLabel: "submit button can't have a variation other than work",
  type: "primary",
  variation: "learning",
  submit: true,
});


// This is new rule enforcement
badButtons.push({
  label: "nav can't go to external links",
  to: "",
  external: true,
});

//done so there isn't an error on badButtons not being used
console.log(badButtons);
```

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

File: packages/components/src/Button/Button.tsx [hunk 1]
```diff
@@ -19,75 +19,73 @@ interface ButtonFoundationProps {
...
-interface SubmitButtonProps
-  extends Omit<ButtonFoundationProps, "external" | "onClick"> {
+interface SubmitButtonProps {
   /**
    * Allows the button to submit a form
    */
-  submit: boolean;
+  submit: true;
+  readonly type?: "primary";
+  readonly variation?: "work";
```
Comment (davidr64) [inline packages/components/src/Button/Button.tsx:74]: It was not obvious before that submit buttons must have the variation "work" however variation could not be added and the component defaults variation to "work" I made this change so `type` and `variation` are consistent.

I have decided not to change existing functionality, though it was discussed [here](https://getjobber.slack.com/archives/C03BJHV3PMG/p1679074908755019?thread_ts=1679073619.972249&cid=C03BJHV3PMG) that we might want to be less restrictive in the future. At which point we would define `type` and `variation` on the ButtonFoundationProps

File: packages/components/src/Button/Button.tsx [hunk 2]
```diff
@@ -19,75 +19,73 @@ interface ButtonFoundationProps {
...
-  BaseActionProps,
-  XOR<DestructiveActionProps, SubmitActionProps>
-> &
-  XOR<SubmitButtonProps, XOR<ButtonLinkProps, ButtonAnchorProps>> &
-  XOR<ButtonIconProps, ButtonLabelProps>;
+/** This type ensures the button defines only one purpose */
+type ButtonFunctionality = XOR<
+  SubmitButtonProps,
+  XOR<ButtonLinkProps, ButtonAnchorProps>
+  >;
```
Comment (davidr64) [inline packages/components/src/Button/Button.tsx:81]: If we added a new type with onClick required we could ensure that all buttons do something.  That seemed like a risky change though and not the purpose of this PR. 
eg
```
interface Clickable  extends BaseActionProps {
    onClick: ()=> void
}
XOR<Clickable, XOR<
  SubmitButtonProps,
  XOR<ButtonLinkProps, ButtonAnchorProps>
  >>```

**Comment by MichaelParadis** [comment]
Suggested edit:
```diff
diff --git a/packages/components/src/Button/Button.tsx b/packages/components/src/Button/Button.tsx
index e87c3fb5..2aef5d54 100644
--- a/packages/components/src/Button/Button.tsx
+++ b/packages/components/src/Button/Button.tsx
@@ -71,7 +71,10 @@ interface SubmitButtonProps {
    */
   submit: true;
   readonly type?: "primary";
-  readonly variation?: "work";
+  readonly variation?: Extract<
+    BaseActionProps["variation"],
+    "work" | "destructive"
+  >;
 }
 
 /** This type ensures the button defines only one purpose */

```

**Review by MichaelParadis — CHANGES_REQUESTED** [review]
Small change see: https://github.com/GetJobber/atlantis/pull/1164#issuecomment-1490556487

**Comment by davidr64** [comment]
@MichaelParadis. To be clear you are asking me to change existing functionality. The submit button did not used to support destructive variation. I am good doing that as long as that is what you wanted.

**Review by MichaelParadis — APPROVED** [review]
LGTM!

**Pull request merged**